### PR TITLE
Pull upstream changes

### PR DESCRIPTION
**Pull request by lachmatt**
## Why

Keep in sync with upstream.

## What

Upstream sync 6989e88ce638ba1e4ee1b178d70e692a3a5499dd..afb96387b2b9c246c8cc29b8285ff944609dcb9b

## Tests

CI

### CODE REVIEW

**Comment by Kielek** [comment]
Notes after short review
tracer\src\Datadog.AutoInstrumentation.NativeLoader\log.h - line #20 - reference to Datadog
ExpandRouteTemplatesEnabled - should it be in unpublisjed or non supported?
DuckType - to string - candidate to update in Otel repository
native code changes - 3rd paird of eyes needed
tracer\test\Datadog.Trace.ClrProfiler.IntegrationTests\NativeProfilerChecks.cs - // TODO: We should figure out why this happens, but hard to reproduce - i it ours or from upstream?
verify what is AutomaticInstrumenation - do you know what is it?
FunctionMethodArgument -> TypeSignature - conflict with #414

File: tracer/src/Datadog.AutoInstrumentation.NativeLoader/log.h [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+#include "../../../shared/src/native-src/string.h"
+
+class Log final
+{
+private:
+    struct NativeLoaderLoggerPolicy
+    {
+        inline static const shared::WSTRING filename = WStr("dotnet-native-loader");
+#ifdef _WIN32
+        inline static const shared::WSTRING folder_path = WStr(R"(Datadog-APM\logs)");
```
Comment (pellared) [inline tracer/src/Datadog.AutoInstrumentation.NativeLoader/log.h:20]: I think this line should be double checked

**Comment by Kielek** [comment]
Answears from @lachmatt 
tracer\src\Datadog.AutoInstrumentation.NativeLoader\log.h - line #20 - reference to Datadog - not used at the moment. Can be changed later.
ExpandRouteTemplatesEnabled - should it be in unpublisjed or non supported? - we can keep where it it. We can decide later if it is not supported.
tracer\test\Datadog.Trace.ClrProfiler.IntegrationTests\NativeProfilerChecks.cs - // TODO: We should figure out why this happens, but hard to reproduce - i it ours or from upstream? - it is from upstream


by me:
DuckType - to string - candidate to update in Otel repository - https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/issues/402
FunctionMethodArgument -> TypeSignature - conflict with #414 - resolved

Still needed
native code changes - 3rd paird of eyes needed
verify what is AutomaticInstrumenation - do you know what is it?

File: tracer/src/Datadog.Trace/Tracer.cs [hunk 2]
```diff
@@ -369,20 +369,19 @@ internal SpanContext CreateSpanContext(ISpanContext parent = null, string servic
                 traceContext = parentSpanContext.TraceContext;
                 if (traceContext == null)
                 {
-                    traceContext = new TraceContext(this);
+                    var traceTags = TraceTagCollection.ParseFromPropagationHeader(parentSpanContext.DatadogTags);
```
Comment (pellared) [inline tracer/src/Datadog.Trace/Tracer.cs:372]: should it be called for SFx?

**Review by pellared — COMMENTED** [review]
Why `docs/development/DuckTyping.md` is removed?

File: tracer/src/Datadog.Trace.Tools.Runner/Checks/GacCheck.cs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            foreach (var folder in Directory.GetDirectories(gacFolder))
+            {
+                if (!File.Exists(Path.Combine(folder, "Datadog.Trace.dll")))
+                {
+                    continue;
+                }
+
+                var name = new DirectoryInfo(folder).Name;
+
+                // Format: v4.0_2.3.0.0__def86d061d0d2eeb
```
Comment (pjanotti) [inline tracer/src/Datadog.Trace.Tools.Runner/Checks/GacCheck.cs:43]: update the public token

Comment (Kielek) [reply tracer/src/Datadog.Trace.Tools.Runner/Checks/GacCheck.cs]: fixed, but IMO whole project should be removed

File: tracer/src/Datadog.AutoInstrumentation.NativeLoader/log.h (hunk 1 above)
Comment (Kielek) [reply tracer/src/Datadog.AutoInstrumentation.NativeLoader/log.h]: Fixed.

File: tracer/src/Datadog.Trace/Tracer.cs (hunk 2 above)
Comment (Kielek) [reply tracer/src/Datadog.Trace/Tracer.cs:372]: Next upstream sync will handle it https://github.com/DataDog/dd-trace-dotnet/pull/2507

**Review by Kielek — APPROVED** [review]

**Pull request merged**